### ISSUE TITLE
random hanging during text editing

### ISSUE THREAD

**Issue by robmcmullen** [robmcmullen/peppy#775]
_[Trac time 20090925 232230Z]_ Occasionally, fundamental mode seems to hang at 100% processor usage for 30 seconds to a minute.  I'm thinking the spell checker might be the cause, but I'm not sure.

**Comment by robmcmullen** [comment]
_[Trac time 20090925 232341Z]_ Trying the python debugger:

```
python
import pdb
import peppy.main
pdb.run('peppy.main.main()')
```

to see if I can use Ctrl-C to break during a hang.

**Comment by robmcmullen** [comment]
_[Trac time 20090925 233610Z]_ Nope, using Ctrl-C in pbd quits the entire thing.

Fortunately, winpdb works to pause the application, so hopefully it will work the next time the problem hits and it is frozen.

**Comment by robmcmullen** [comment]
_[Trac time 20090926 190948Z]_ Looks like it might be in processDirtyRanges.  When deleting a bunch of text using the del key, I seem to be able to trigger the probem.

processDirtyRanges calls checkRange with the current word start and current word end.  Maybe those values are bogus?

**Comment by robmcmullen** [comment]
_[Trac time 20090926 193243Z]_ Yep, `checkRange` is being called with values that cause the count to be -1

Can regularly duplicate it now by holding down the delete key in a spell checking region, and using the mouse to select regions after the cursor.  When those regions are deleted, the bug is triggered.

**Comment by robmcmullen** [comment]
_[Trac time 20090926 194848Z]_ f9f62cbc7fb68d9c6401cc2886db65b2b44d1566 Fixed #775: added boundary check to make sure word end doesn't come before word start
- when deleting lots of text, it was possible for the current word end to move before the word start
- this triggered the call to SetStyling to have a negative number for the number of characters, causing the freeze
